lib/Dialect/RTL/RTLFolds.cpp

```diff
@@ -706,3 +714,259 @@ void ConcatOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
+    return lhs.sle(rhs);
+  case ICmpPredicate::sgt:
+    return lhs.sgt(rhs);
+  case ICmpPredicate::sge:
+    return lhs.sge(rhs);
+  case ICmpPredicate::ult:
+    return lhs.ult(rhs);
+  case ICmpPredicate::ule:
+    return lhs.ule(rhs);
+  case ICmpPredicate::ugt:
+    return lhs.ugt(rhs);
+  case ICmpPredicate::uge:
+    return lhs.uge(rhs);
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+// Returns the result of applying the predicate when the LHS and RHS are the
+// exact same value.
+static bool applyCmpPredicateToEqualOperands(ICmpPredicate predicate) {
+  switch (predicate) {
+  case ICmpPredicate::eq:
+  case ICmpPredicate::sle:
+  case ICmpPredicate::sge:
+  case ICmpPredicate::ule:
+  case ICmpPredicate::uge:
+    return true;
+  case ICmpPredicate::ne:
+  case ICmpPredicate::slt:
+  case ICmpPredicate::sgt:
+  case ICmpPredicate::ult:
+  case ICmpPredicate::ugt:
+    return false;
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+// Try to apply the predicate to an unknown LHS but with a constant RHS.
+// This is accomplished by reasoning about RHS which are on the upper or
+// lower boundary of possible values.
+static IntegerAttr tryApplyCmpPredicateToRHS(ICmpPredicate predicate,
+                                             IntegerAttr &rhs) {
+  MLIRContext *context = rhs.getContext();
+  auto value = rhs.getValue();
+  switch (predicate) {
+  case ICmpPredicate::slt:
+    if (value.isMinSignedValue())
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::sle:
+    if (value.isMaxSignedValue())
+      return getTrueAttr(context);
+    return {};
+  case ICmpPredicate::sgt:
+    if (value.isMaxSignedValue())
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::sge:
+    if (value.isMinSignedValue())
+      return getTrueAttr(context);
+    return {};
+  case ICmpPredicate::ult:
+    if (value.isMinValue())
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::ule:
+    if (value.isMaxValue())
+      return getTrueAttr(context);
+    return {};
+  case ICmpPredicate::ugt:
+    if (value.isMaxValue())
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::uge:
+    if (value.isMinValue())
+      return getTrueAttr(context);
+    return {};
+  default:
+    return {};
+  }
+}
+
+OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
+
+  // gt a, a -> false
+  // gte a, a -> true
+  if (lhs() == rhs()) {
+    auto val = applyCmpPredicateToEqualOperands(predicate());
+    return IntegerAttr::get(getType(), val);
+  }
+
+  auto lhs = constants[0].dyn_cast_or_null<IntegerAttr>();
+  auto rhs = constants[1].dyn_cast_or_null<IntegerAttr>();
+
+  // gt 1, 2 -> false
+  if (lhs && rhs) {
+    auto val = applyCmpPredicate(predicate(), lhs.getValue(), rhs.getValue());
+    return IntegerAttr::get(getType(), val);
+  }
+
+  // lt x, 0 -> false
+  if (rhs)
+    return tryApplyCmpPredicateToRHS(predicate(), rhs);
```

Comment:
Ok, I know I talked you out of this, but if you're going to handle the "constant on the RHS in fold" at all, then it does make sense to handle "constant on the LHS" as well like you were doing.  I was suggesting you drop this entirely so you didn't need to define `tryApplyCmpPredicateToRHS`

Reply:
I missed the part that it should be handled totally in canonicalization. Woops!